src/transformers/tokenization_utils_base.py

```diff
@@ -2231,36 +2191,59 @@ def _from_pretrained(
+                        else:
+                            special_tokens.append(str(init_kwargs[key]))
+
                 with open(added_tokens_file, encoding="utf-8") as added_tokens_handle:
                     added_tok_encoder = json.load(added_tokens_handle)
-                # legacy: we have to init with (rstrip=True, lstrip=True)
-                strip = True if "Fast" not in cls.__name__ else False
-                added_tokens_decoder = {
-                    index: AddedToken(token, rstrip=strip, lstrip=strip) for token, index in added_tok_encoder.items()
-                }
+                for str_token, index in added_tok_encoder.items():
+                    # if index not in added_tokens_decoder and str_token not in added_tokens_map:
+                    special = str_token in special_tokens
+                    added_tokens_decoder[index] = AddedToken(
+                        str_token, rstrip=False, lstrip=False, normalized=not special, special=special
+                    )
+                    added_tokens_map[str(token)] = added_tokens_decoder[index]
+
+            # allows converting a fast -> slow: add the `tokenizer.json`'s `"added_tokens"` to the slow tokenizer
+            # if `tokenizer_config.json` is `None`
+            if "Fast" not in cls.__name__ and tokenizer_file is not None:
+                # This is for slow so can be done before
+                with open(tokenizer_file, encoding="utf-8") as tokenizer_file_handle:
+                    tokenizer_file_handle = json.load(tokenizer_file_handle)
+                    added_tokens = tokenizer_file_handle.pop("added_tokens")
+                for serialized_tokens in added_tokens:
+                    idx = serialized_tokens.pop("id")
+                    added_tokens_decoder[idx] = AddedToken(**serialized_tokens)
+                    added_tokens_map[str(added_tokens_decoder[idx])] = added_tokens_decoder[idx]
             # end legacy
 
-        # slow -> fast, non-legacy: we need to make sure the `added_tokens_decoder` is used to add tokens if the `fast` was not properly saved!
-        # thus we delay adding special tokens in the init using `slow_to_fast` flag.
-        if added_tokens_decoder is not {} and "Fast" in cls.__name__:
-            init_kwargs["slow_to_fast"] = True
-        if len(additional_special_tokens) > 0:
-            init_kwargs["additional_special_tokens"] = additional_special_tokens
-        init_kwargs["added_tokens_decoder"] = added_tokens_decoder
+        # Passing AddedTokens and not strings to the class to prevent it from casting the string to a different AddedToken
+        for key in cls.SPECIAL_TOKENS_ATTRIBUTES & init_kwargs.keys():
+            if added_tokens_map != {} and init_kwargs[key] is not None:
+                if key != "additional_special_tokens":
+                    init_kwargs[key] = added_tokens_map.get(init_kwargs[key], init_kwargs[key])
```

Comment:
Hi @ArthurZucker,

I think this is supposed to be 
`init_kwargs[key] = added_tokens_map.get(key, init_kwargs[key])`
As `init_kwargs[key]` will return a dict in some cases, and cause unhashable type error:
```
site-packages/transformers/tokenization_utils_base.py", line 2242, in _from_pretrained
    init_kwargs[key] = added_tokens_map.get(init_kwargs[key], init_kwargs[key])
TypeError: unhashable type: 'dict'
```
Thanks.